### PR TITLE
Consistency fix with Fastly: querysting fixed to URL encode keys and …

### PR DESCRIPTION
**Pull request by akrainiouk**
…encode space with '%20' rather then '+'

falco implementation of querystring does not urlencode query parameter names which can produce invalid urls
Also falco implementation encodes space character with '+' which is inconsistent with Fastly. Fastly use %20 for space encoding.

### CODE REVIEW

**Review by richardmarshall — CHANGES_REQUESTED** [review]
Thanks. Fixing consistency issues with Fastly behavior is appreciated.

File: interpreter/function/shared/querystring.go [hunk 1]
```diff
@@ -142,9 +142,9 @@ func (q *QueryStrings) String() string {
 			buf.WriteString(key)
 		} else {
 			for j := range v.Value {
-				buf.WriteString(key)
+				buf.WriteString(url.PathEscape(key))
 				buf.WriteString("=")
-				buf.WriteString(url.QueryEscape(v.Value[j]))
+				buf.WriteString(url.PathEscape(v.Value[j]))
```
[the remark this answers is not in the document]
Comment (akrainiouk) [reply interpreter/function/shared/querystring.go]: Oops! Totally forgot about query specifics. Thanks for catching!
Following your advise created a wrapper that first uses urlQueryEscape and then replaces '+' occurrences with with "%20". This should be safe as original '+' characters will be encoded with "%2B".
I also added unit tests that compare encoding results with the values taken directly from Fastly execution.

**Comment by richardmarshall** [comment]
Did a bit of looking at the details of how Fastly handles query escapes with the `querystring.*` functions. From a non exhaustive check it seems like they're performing escaping only when setting values but not when reading them or performing other operations on them (filtering, sorting).

For example the following will log `/?a=b+c&b=c%20d` not `/?a=b%20c&b%20c`.

```vcl
log querystring.set("/?a=b+c", "b", "c d");
```

Fiddle with a few more examples: https://fiddle.fastly.dev/fiddle/e019c129

If you're up to tackling investigating this further and incorporating the needed changes to the PR that would be great. If not we can create an issue to track the remaining work.

**Comment by akrainiouk** [comment]
Hi Richard,
I am afraid we are opening a Pandora box :)
The problem of course is deeper than I attempted to address here. Current implementation first parses the source value into a structure then manipulates it and finally converts back to string. This however introduces several unwanted side effects such as grouping name value pairs (implied by underlying query structure), it also makes it hard to follow various Fastly quirks. For instance in some cases (e.g. sort) original '?' separator is preserved even if the query is empty, in other cases (filter) it is removed.
Another quirk is that sorting is applied to unsplit name=value pairs so that
```
querystring.sort({"?foo=one&foo>=too&foo<=three"})
```
 returns 
```
?foo<=three&foo=one&foo>=too
```
 and not
```
?foo=one&foo<=three&foo>=two
```
 as one might expect.
If you guys are interested, in [this PR draft](https://github.com/ysugimoto/falco/pull/264) I provided an alternative implementation of querystring functions that do not try to pre-parse the source string and attempt to follow Fastly behavior as much as I can check.
I converted existing unit tests so that at least existing test cases are still valid and added some more (the behavior is more of less covered in [this Fastly fiddle](https://fiddle.fastly.dev/fiddle/496f0b10).

**Comment by akrainiouk** [comment]
Closing it for now as this PR was sitting dormant for a long time
